rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+# Continuous Deployment
+
+## Summary
+
+In [RFC 70](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-070-path-towards-continuous-deployment-cd.md#proposal) we committed to fully automatic deployments. The following provides a framework for how we will actually achieve this for all of our existing apps.
+
+Decision: in order to safely enable automatic deployments for each app, we will specify minimum criteria that must first be met. See the criteria described [below](#safety-concerns).
+
+An audit of our apps against these criteria can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+## Problem
```

Comment:
Having attended your talk, I now have a much clearer understanding of the intent behind the RFC. I do think it's worth clarifying the _Problem_ section in light of what you said in the talk, as I got a very different impression as to the goals of the RFC from reading this section than I did from the talk.

My understanding from the talk is that the problem is to determine a set of minimum criteria for enabling CD for an app such that we can be reasonably confident that enabling CD won't significantly increase the rate of defects which reach production. This is currently mentioned briefly in the summary section above, but it wasn't at all clear to me that that's the whole purpose of the RFC.

My understanding from reading this Problem section was much more vague - it seemed to be partly setting out a case for why we want continuous deployment (which, as you mentioned in the talk, has already been covered elsewhere) and partly saying that GOV.UK has problems with testing (which is true and relevant but if I understood you correctly in the talk, the test engineering problems aren't really what you're trying to solve here - it's just that the current lack of testing makes it more challenging to say what a reasonable minimum standard should be without creating unrealistic hurdles to enabling CD).

I'd recommend basically replacing this section with a problem statement about determining those minimum criteria (basically a more in-depth version of the above in your own words - assuming I've not misinterpreted you). I think most of what's currently in this section is either unrelated or only peripherally related to the problem which you're setting out to solve here.

Reply:
Thanks for taking the time to get all this down. It makes a lot of sense, and you're not the only person who seems to be confused about the purpose of this RFC. I'm open to making changes here.

I've found the existence of RFC 70 rather awkward. It barely justifies why CD is the way to go, and goes into even less detail about how we'd actually get there. I decided to basically ignore it when writing this one.

Part of the original goal for this RFC was to act as a writeup for the trials we've been doing on CD over the past few months. I've filled in a lot of the holes left behind by RFC 70: the formal justification, the mechanism, the security restrictions. While the safety criteria are the key difference between the trials and the change proposed here, I still think we should document the overall CD approach as a clear decision (that's not in a Google Doc).

I'm not sure where to go here, as I see pros/cons either way. One simple thing that might help: what did you think the purpose of the RFC was originally? What impact did that have on your reading of it?

> as I got a very different impression as to the goals of the RFC from reading this section than I did from the talk

I may not have helped by deliberately down-playing the rest of the RFC. I only did this, though, because I knew most of the people present had been involved with the trials and previous discussions of CD.

Reply:
Ah I see - if the remit of this RFC _does_ include making the case for CD then I'd recommend organising those "why CD" arguments under a section heading so that's it's clear to the reader and dealt with separately from the "where do we set the bar for testing before CDing an app" discussion.

Reply:
I'm thinking about what this change might look like - ideally, something simple. How about this (outline):

## Problem

Our current approach to deployment has multiple issues with it:

### 1. We have a constant backlog of deployments
**[move existing problem intro here]**

### 2. We do not have a working definition of "OK"
**[append the following]**

While we often rely on automated tests to tell us when something is broken, it's unclear whether this is justified: there is no minimum standard for testing of GOV.UK apps.

### 3. We freely merge PRs with breaking changes

## Proposal

The obvious solution to the first issue is to make deployments automatic for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145). This has already begun, as part of [a trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).

> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].

However, we have been unable to make further progress due to the other two issues: we are concerned that enabling automatic deployments en masse would lead to an increase in defects reaching production.

In order to proceed, we need to agree a clear and actionable set of criteria for enabling automatic deployments, so that we will only do so when we can show this is safe, as well as secure.

### Security checks

### Safety checks

Reply:
Cool - yes I think that's much clearer. Thanks!

Reply:
Great! I've put this in a new commit (with a few extra edits): https://github.com/alphagov/govuk-rfcs/pull/128/commits/a593468a87998575690dfb06ff61ee827c6c890f.

Feel free to comment after reading, if you think there's more to discuss.